File: problem-solving/_lc-5-medium-longest-palindromic-substring/1-solution.py
"""
https://leetcode.com/problems/longest-palindromic-substring/

Complexity
----------
Time: O(n^2)
Space: O(1)
"""

class Solution:
    def isPalindrome(self, s, l, r, start, end):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            if end - start < r - l:
                start, end = l, r
            l -= 1
            r += 1
        return start, end

    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        start = end = 0
        for i in range(n):
            # Odd len palindrome - having center char at i
            start, end = self.isPalindrome(s, i - 1, i + 1, start, end)

            # Even len palindrome - don't have center char
            start, end = self.isPalindrome(s, i, i + 1, start, end)
        return s[start:end + 1]

File: problem-solving/_lc-5-medium-longest-palindromic-substring/1-solution.go
/*
https://leetcode.com/problems/longest-palindromic-substring/

Complexity
----------
Time: O(n^2)
Space: O(1)
*/

func longestPalindrome(s string) string {
    start, end := 0, 0
    n := len(s)
    for i := 0; i < n; i++ {
        // Odd len palindrome - having center char
        l := i - 1
        r := i + 1
        for l >= 0 && r < n && s[l] == s[r] {
            if end - start < r - l {
                start, end = l, r
            }
            l -= 1
            r += 1
        }
        // Even len palindrome - don't have center char
        l = i
        r = i + 1
        for l >= 0 && r < n && s[l] == s[r] {
            if end - start < r - l {
                start, end = l, r
            }
            l -= 1
            r += 1
        }
    }
    return s[start:end + 1]
}
